designs/mfa_backup_code.rst

```diff
@@ -0,0 +1,81 @@
+
+Account recovery (backup) mfa codes draft design
+------------------------
+MFA requires users to have access to a working physical device, however that might not be always possible.
+
+Backup mfa code allows:
+
+- single-device accounts to be restored
+- initial setup for passwordless-login
+
+Overview
+------------------------
+
+If the user has a Backup code, then they can replace the TOTP/WebAuthn challenge with it during login. For example, TOTP + password auth will be Backup code + password instead.
+
+To implement this, we can add Backup code as a new AuthCredential, so that it can be used in the auth process.
+
+To validate Backup code, we simply check if the input matches the Backup code saved in database, which should be hashed with salt (similar to how password is validated).
+
+A user can only have one Backup code, if user wants a new Backup code:
+
+- if the existing Backup code is still valid, cannot create new Backup code
+- if no existing/valid Backup code, create/replace with a new Backup code
+
+The login session should only last for 5 mins with a Claim that allows users to change login credentials.
+
+Code Generation
+------------------------
+The Backup code could be generated by TOTP/HOTP, or a cryptographically secure pseudorandom number generator. 
+Attackers should not be able to guess it even if they know our source code.
+
+Implementation 1
+------------------------
+
+Based on the Microsoft temporary access pass:
+
+Add a new class/object: Backup code policy
+
+- Each Kanidm has a Backup code policy
+- Define which users/groups can use the Backup code
+   - User can only login if they are included in the Backup code policy
+- Define parameters like max/min lifetime, one-time use and length
+- Only ``idm_hp_account_manage_priv`` can modify it?
```

Comment:
Or self modification if they are in the correct policy group. I think this policy actually could already be done pretty easily via a membership to a backup_code_allowed group and the existing access profile infrastructure.